File: day10.pl
#!/usr/bin/env perl 
#===============================================================================
#
#         FILE: day10.pl
#
#        USAGE: ./day10.pl  
#
#  DESCRIPTION: Day 10 - Elves play look-and-say
#
#      OPTIONS: ---
# REQUIREMENTS: ---
#         BUGS: ---
#        NOTES: Part 1 (loop limit set to 40) runs in reasonable time but for
#               part 2 (loop limit set to 50) it takes well over an hour to
#               run.  I expect that any substantial reduction in runtime would
#               require not using regular expressions.
#       AUTHOR: Mark T. Johnson (mtj), markj@matzsoft.com
# ORGANIZATION: MATZ Software & Consulting
#      VERSION: 1.0
#      CREATED: 12/10/15 08:58:16
#     REVISION: ---
#===============================================================================

use strict;
use warnings;
use utf8;
use 5.010;


my $input = 1113122113;
my $current = $input;

for ( 1 .. 50 ) {
    my $next;

    while ( my $match = $current =~ m/(\d)\1*/pg ) {
        $next .= length( ${^MATCH} ) . $1;
    }

    $current = $next;
}

say length( $current );

File: 2015/day10.swift
//
//         FILE: main.swift
//  DESCRIPTION: day10 - Elves Look, Elves Say
//        NOTES: ---
//       AUTHOR: Mark T. Johnson, markj@matzsoft.com
//    COPYRIGHT: © 2021 MATZ Software & Consulting. All rights reserved.
//      VERSION: 1.0
//      CREATED: 07/05/21 22:45:24
//

import Foundation
import Library

func rounds( initial: String, rounds: Int ) -> String {
    var current = initial
    
    for _ in 1 ... rounds {
        var last: Character = "1"
        var count = 0

        current = current.map { character -> String in
            if character == last {
                count += 1
                return ""
            }
            
            let retval = count == 0 ? "" : "\(count)\(last)"
            last = character
            count = 1
            return retval
        }.joined()
        if count > 0 {
            current += "\(count)\(last)"
        }
    }
    
    return current
}


func part1( input: AOCinput ) -> String {
    let result = rounds( initial: input.line, rounds: 40 )
    return "\( result.count )"
}


func part2( input: AOCinput ) -> String {
    let result = rounds( initial: input.line, rounds: 50 )
    return "\( result.count )"
}


try print( projectInfo() )
try runTests( part1: part1 )
try runTests( part2: part2 )
try solve( part1: part1 )
try solve( part2: part2 )
